spring-integration-jdbc/src/test/java/org/springframework/integration/jdbc/postgres/PostgresChannelMessageTableSubscriberTests.java

```diff
@@ -310,6 +311,41 @@ public void testRenewConnection() throws Exception {
 		assertThat(payloads).containsExactlyInAnyOrder("1", "2");
 	}
 
+	@Test
+	public void testUnsubscribeHandlerDuringDispatch() throws InterruptedException {
+		int numberOfMessages = 150;
+		CountDownLatch messagesUntilUnsubscribe = new CountDownLatch(10);
```

Comment:
Why wait for just 10 if we produce so many?
Maybe then decrease a production amount?

Reply:
Well, the first `10` should ensure that processing started. If I reduce the overall number of messages, the test is not stable on my system, as the processing is just too fast. Adding a second subscriber even increases that number to 200, as the `unsubscribe` of both `MessageHandler`s take a bit longer now.

Reply:
> take a bit longer now

Does your `fair` comment before make any difference?
I mean it really would be fair to give an `unsubscribe` (`WriteLock`) a precedence to shutdown quicker, rather than let all those reading threads to poll more messages.

I think if you decrease the number of reading threads , that would make some difference on the test execution.

However, I'm leaving this comment just for a curiosity.
Looks like we have reached a consensus when our solution works as expected and test does not take too much time.

Let me know if you are going to look into that or we can merge your solution already now!

Reply:
I didn't check it before, but a rough measurement (simply executing the test a few times) shows no observable difference between a normal and a `fair` `ReentrantReadWriteLock`. So, I would leave it as is.

The overall number of messages does not really matter and could even be higher, as processing is interrupted by unsubscribing the handlers anyway. The runtime of "a normal test execution" is defined by the time the `unsubscribe`s take. As the test seems stable, I would leave it this way.